packages/@glimmer/validator/lib/tracking.ts

```diff
@@ -72,88 +81,145 @@ export function endTrackFrame(): Tag {
+}
 
-    return lastValue!;
+//////////
+
+// public interface
+export interface Cache<T = unknown> {
+  [CACHE_KEY]: T;
+}
+
+const FN: unique symbol = symbol('FN');
+const LAST_VALUE: unique symbol = symbol('LAST_VALUE');
+const TAG: unique symbol = symbol('TAG');
+const SNAPSHOT: unique symbol = symbol('SNAPSHOT');
+const DEBUG_LABEL: unique symbol = symbol('DEBUG_LABEL');
+
+interface InternalCache<T = unknown> {
+  [FN]: (...args: unknown[]) => T;
+  [LAST_VALUE]: T | undefined;
+  [TAG]: Tag | undefined;
+  [SNAPSHOT]: Revision;
+  [DEBUG_LABEL]?: string | false;
+}
+
+export function createCache<T>(fn: () => T, debuggingLabel?: string | false): Cache<T> {
+  if (DEBUG && !(typeof fn === 'function')) {
+    throw new Error(
+      `createCache() must be passed a function as its first parameter. Called with: ${String(fn)}`
+    );
+  }
+
+  let cache: InternalCache<T> = {
+    [FN]: fn,
+    [LAST_VALUE]: undefined,
+    [TAG]: undefined,
+    [SNAPSHOT]: -1,
   };
 
   if (DEBUG) {
-    IS_CONST_MAP.set(memoized, undefined);
+    cache[DEBUG_LABEL] = debuggingLabel;
   }
 
-  return memoized;
+  return (cache as unknown) as Cache<T>;
 }
 
-export function isConstMemo(memoized: Function) {
-  if (DEBUG && IS_CONST_MAP.has(memoized) && IS_CONST_MAP.get(memoized) === undefined) {
+export function getValue<T>(cache: Cache<T>): T {
+  if (!isCache(cache)) {
+    throw new Error(
+      `getValue() can only be used on an instance of a cache created with createCache(). Called with: ${String(
+        cache
+      )}`
+    );
+  }
+
+  let fn = cache[FN];
+  let tag = cache[TAG];
+  let snapshot = cache[SNAPSHOT];
+
+  if (tag === undefined || !validateTag(tag, snapshot)) {
+    beginTrackFrame();
+
+    try {
+      if (DEBUG) {
+        runInAutotrackingTransaction!(() => (cache[LAST_VALUE] = fn()), cache[DEBUG_LABEL]);
+      } else {
+        cache[LAST_VALUE] = fn();
+      }
+    } finally {
+      tag = endTrackFrame();
+      cache[TAG] = tag;
+      cache[SNAPSHOT] = valueForTag(tag);
+      consumeTag(tag);
+    }
+  } else {
+    consumeTag(tag);
+  }
+
+  return cache[LAST_VALUE]!;
+}
+
+export function isConst(cache: Cache) {
+  if (!isCache(cache)) {
+    throw new Error(
+      `isConst() can only be used on an instance of a cache created with createCache(). Called with: ${String(
+        cache
+      )}`
+    );
+  }
+
+  if (DEBUG && cache[TAG] === undefined) {
     throw new Error(
-      'Attempted to call `isConstMemo` on a memoized function, but the function has not been run at least once yet. You cannot know if a memoized function is constant or not until it has been run at least once. Call the function, then pass it to `isConstMemo`.'
+      `isConst() can only be used on a cache once getValue() has been called at least once. Called with cache function:\n\n${String(
+        cache[FN]
+      )}`
     );
   }
 
-  return IS_CONST_MAP.get(memoized) === true;
+  return isConstTag(cache[TAG]!);
```

Comment:
you can put the debug error inside a type assertion function like isCache to remove the !